README
Bolo-Socket

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Bolo::Socket

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bolo-Socket

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Bolo-Socket

    CPAN Ratings
        http://cpanratings.perl.org/d/Bolo-Socket

    Search CPAN
        http://search.cpan.org/dist/Bolo-Socket/


LICENSE AND COPYRIGHT

Copyright (C) 2016 Dan Molik

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Bolo-Socket

0.01    Date/time
        First version, released on an unsuspecting world.



lib/Bolo/Socket/ZMQ.pm
package Bolo::Socket::ZMQ;

use strict;
use warnings;

use ZMQ::LibZMQ3;
use ZMQ::Constants qw/
	ZMQ_PUB ZMQ_SUB ZMQ_DEALER ZMQ_ROUTER
	ZMQ_SNDMORE ZMQ_RCVMORE
	ZMQ_SUBSCRIBE
	ZMQ_LINGER
	ZMQ_POLLIN ZMQ_POLLOUT
/;

sub new
{
	my ($class, %opts) = @_;
	my $ctx = zmq_ctx_new()
		or return undef;

	bless({
		_context   => $ctx,
		_timeout   => $opts{timeout} || 5,
		_endpoints => [],
		_binds     => [],
		_socket    => undef,
	}, $class);
}

sub timeout
{
	my ($self, $val) = @_;
	return $self->{_timeout} unless defined $val;

	my $old = $self->{_timeout};
	$self->{_timeout} = $val;
	return $old;
}

sub timeout_us
{
	my ($self) = @_;
	my $timeout = $self->timeout;
	return -1 if $timeout <= 0;

	$timeout * 1000; # s -> us
}

sub _socket
{
	my ($self, $type) = @_;
	die "Socket type already set\n" if $self->{_socket};

	$self->{_socket} = zmq_socket($self->{_context}, $type)
		or die "Failed to create a socket: $!\n";
	$self->{type} = $type;

	zmq_setsockopt($self->{_socket}, ZMQ_LINGER, 0) == 0
		or die "Failed to set 0MQ linger timeout to 0\n";

	$self;
}

sub router { $_[0]->_socket(ZMQ_ROUTER); }
sub dealer { $_[0]->_socket(ZMQ_DEALER); }

sub publisher { $_[0]->_socket(ZMQ_PUB); }
sub subscriber
{
	my ($self, $filter) = @_;
	$self->_socket(ZMQ_SUB);

	zmq_setsockopt($self->{_socket}, ZMQ_SUBSCRIBE, $filter || "") == 0
		or die "Failed to set 0MQ subscriber filter\n";
	$self;
}

sub bound
{
	my ($self) = @_;
	return scalar(@{ $self->{_binds} }) > 0;
}

sub bind
{
	my ($self, @listen) = @_;
	die "bind() before setting up socket type (by ->publisher or ->router)\n"
		if !$self->{_socket};

	for (@listen) {
		zmq_bind($self->{_socket}, $_) == 0
			or die "Failed to bind to $_: $!\n";

		push @{ $self->{_binds} }, $_;
	}
	$self;
}

sub connected
{
	my ($self) = @_;
	return scalar(@{ $self->{_endpoints} }) > 0;
}

sub connect
{
	my ($self, @endpoints) = @_;
	die "connect() before setting up socket type (by ->subscriber or ->dealer)\n"
		if !$self->{_socket};

	for (@endpoints) {
		zmq_connect($self->{_socket}, $_) == 0
			or die "Failed to connect to $_: $!\n";
		push @{ $self->{_endpoints} }, $_;
	}
	$self;
}

sub disconnect
{
	my ($self, @endpoints) = @_;
	return $self unless $self->{_socket};

	if (!@endpoints) {
		@endpoints = @{ $self->{_endpoints} };
	}
	for (@endpoints) {
		zmq_disconnect($self->{_socket}, $_);
	}
	$self;
}

sub receive
{
	my ($self) = @_;
	return undef unless $self->{_socket};

	my ($buf, @frames);
	my @ready = zmq_poll([
			{ socket   => $self->{_socket},
			  events   => ZMQ_POLLIN,
			  callback => sub {}   # dummy callback, required by ZMQ::LibZMQ3 impl.
			},
		], $self->timeout_us);
	return undef unless $ready[0];

	do {
		zmq_recv($self->{_socket}, $buf, 4096);
		$buf =~ s/\0*$//;
		push @frames, "$buf";
	} while (zmq_getsockopt($self->{_socket}, ZMQ_RCVMORE));

	if ($self->{type} == ZMQ_ROUTER) {
		my $id = shift @frames;
		shift @frames;
		unshift @frames, $id;
	}

	print STDERR "received ZMQ frames:\n".Dumper(\@frames)
		if $ENV{BOLO_WEB_DEBUG};

	return \@frames;
}

sub send
{
	my ($self, $frames) = @_;
	return undef unless $self->{_socket};

	print STDERR "sending ZMQ frames:\n".Dumper($frames)
		if $ENV{BOLO_WEB_DEBUG};

	my @ready = zmq_poll([
			{ socket   => $self->{_socket},
			  events   => ZMQ_POLLOUT,
			  callback => sub {}   # dummy callback, required by ZMQ::LibZMQ3 impl.
			},
		], $self->timeout_us);
	return undef unless $ready[0];

	if ($self->{type} == ZMQ_DEALER) {
		zmq_send($self->{_socket}, "", 0, ZMQ_SNDMORE) > -1
			or return 0;
	}

	$frames = [map { $_ ? $_ : '' } @$frames];
	my $last = pop @$frames;
	for (@$frames) {
		zmq_send($self->{_socket}, $_, -1, ZMQ_SNDMORE) > -1
			or return 0;
	}
	zmq_send($self->{_socket}, $last, -1) > -1
		or return 0;
	return 1;
}

sub shutdown
{
	my ($self) = @_;
	return unless $self->{_socket};

	if ($self->bound) {
		zmq_unbind($self->{_socket}, $_)
			for @{ $self->{_binds} };
		$self->{_binds} = [];
	}
	if ($self->connected) {
		zmq_disconnect($self->{_socket}, $_)
			for @{ $self->{_endpoints} };
		$self->{_endpoints} = [];
	}

	zmq_close($self->{_socket}) == 0
		or die "Failed to close socket: $!\n";
	$self->{_socket} = undef;
}

sub DESTROY
{
	my ($self) = @_;
	$self->shutdown;
	zmq_ctx_destroy($self->{_context});
}

1;


lib/Bolo/Socket.pm
package Bolo::Socket;

use strict;
use warnings;

use base qw/Exporter/;

our $VERSION = '0.02';

use ZMQ::LibZMQ3;
use ZMQ::Constants qw/
	ZMQ_PUB ZMQ_SUB ZMQ_DEALER ZMQ_ROUTER
	ZMQ_PUSH ZMQ_PULL
	ZMQ_SNDMORE ZMQ_RCVMORE
	ZMQ_SUBSCRIBE
	ZMQ_LINGER
	ZMQ_POLLIN ZMQ_POLLOUT
/;

use constant {
	OK       => 0,
	WARNING  => 1,
	CRITICAL => 2,
	UNKNOWN  => 3
};

our @EXPORT = qw/
	submit_metric submit_state
	OK WARNING CRITICAL UNKNOWN
/;

sub new
{
	my ($class, %opts) = @_;
	my $ctx = zmq_ctx_new(3)
		or return undef;

	bless({
		_context   => $ctx,
		_timeout   => $opts{timeout} || 5,
		_endpoints => [],
		_binds     => [],
		_socket    => undef,
	}, $class);
}

sub timeout
{
	my ($self, $val) = @_;
	return $self->{_timeout} unless defined $val;

	my $old = $self->{_timeout};
	$self->{_timeout} = $val;
	return $old;
}

sub timeout_us
{
	my ($self) = @_;
	my $timeout = $self->timeout;
	return -1 if $timeout <= 0;

	$timeout * 1000; # s -> us
}

sub _socket
{
	my ($self, $type) = @_;
	die "Socket type already set\n" if $self->{_socket};

	$self->{_socket} = zmq_socket($self->{_context}, $type)
		or die "Failed to create a socket: $!\n";
	$self->{type} = $type;

	#zmq_setsockopt($self->{_socket}, ZMQ_LINGER, 0) == 0
	#	or die "Failed to set 0MQ linger timeout to 0\n";

	$self;
}

sub router { $_[0]->_socket(ZMQ_ROUTER); }
sub dealer { $_[0]->_socket(ZMQ_DEALER); }
sub pusher { $_[0]->_socket(ZMQ_PUSH);   }
sub puller { $_[0]->_socket(ZMQ_PULL);   }
sub publisher { $_[0]->_socket(ZMQ_PUB); }
sub subscriber
{
	my ($self, $filter) = @_;
	$self->_socket(ZMQ_SUB);

	zmq_setsockopt($self->{_socket}, ZMQ_SUBSCRIBE, $filter || "") == 0
		or die "Failed to set 0MQ subscriber filter\n";
	$self;
}

sub bound
{
	my ($self) = @_;
	return scalar(@{ $self->{_binds} }) > 0;
}

sub bind
{
	my ($self, @listen) = @_;
	die "bind() before setting up socket type (by ->publisher or ->router)\n"
		if !$self->{_socket};

	for (@listen) {
		zmq_bind($self->{_socket}, $_) == 0
			or die "Failed to bind to $_: $!\n";

		push @{ $self->{_binds} }, $_;
	}
	$self;
}

sub connected
{
	my ($self) = @_;
	return scalar(@{ $self->{_endpoints} }) > 0;
}

sub connect
{
	my ($self, @endpoints) = @_;
	die "connect() before setting up socket type (by ->subscriber or ->dealer)\n"
		if !$self->{_socket};

	for (@endpoints) {
		zmq_connect($self->{_socket}, $_) == 0
			or die "Failed to connect to $_: $!\n";
		push @{ $self->{_endpoints} }, $_;
	}
	$self;
}

sub disconnect
{
	my ($self, @endpoints) = @_;
	return $self unless $self->{_socket};

	if (!@endpoints) {
		@endpoints = @{ $self->{_endpoints} };
	}
	for (@endpoints) {
		zmq_disconnect($self->{_socket}, $_);
	}
	$self;
}

sub receive
{
	my ($self) = @_;
	return undef unless $self->{_socket};

	my ($buf, @frames);
	my @ready = zmq_poll([
			{ socket   => $self->{_socket},
			  events   => ZMQ_POLLIN,
			  callback => sub {}   # dummy callback, required by ZMQ::LibZMQ3 impl.
			},
		], $self->timeout_us);
	return undef unless $ready[0];

	do {
		zmq_recv($self->{_socket}, $buf, 4096);
		$buf =~ s/\0*$//;
		push @frames, "$buf";
	} while (zmq_getsockopt($self->{_socket}, ZMQ_RCVMORE));

	if ($self->{type} == ZMQ_ROUTER) {
		my $id = shift @frames;
		shift @frames;
		unshift @frames, $id;
	}

	return \@frames;
}

sub send
{
	my ($self, $frames) = @_;
	return undef unless $self->{_socket};

	my @ready = zmq_poll([
			{ socket   => $self->{_socket},
			  events   => ZMQ_POLLOUT,
			  callback => sub {}   # dummy callback, required by ZMQ::LibZMQ3 impl.
			},
		], $self->timeout_us);
	return undef unless $ready[0];

	if ($self->{type} == ZMQ_DEALER || $self->{type} == ZMQ_PUSH) {
		zmq_send($self->{_socket}, "", 0, ZMQ_SNDMORE) > -1
			or return 0;
	}

	$frames = [map { $_ ? $_ : '' } @$frames];
	my $last = pop @$frames;
	for (@$frames) {
		zmq_send($self->{_socket}, $_, -1, ZMQ_SNDMORE) > -1
			or return 0;
	}
	zmq_send($self->{_socket}, $last, -1) > -1
		or return 0;
	return 1;
}

sub submit_metric
{
	my ($type, $name, $value, $endpoint, $time) = @_;
	$type = uc($type);
	return 0 unless $type =~ m/RATE|COUNTER|SAMPLE/;
	my $pdu = [$type, $time || time, $name, $value];
	my $push = Bolo::Socket->new->pusher;
	$push->connect($endpoint);
	$push->send($pdu);
	$push->disconnect($endpoint);
}

sub submit_state
{
	my ($code, $name, $value, $endpoint, $time) = @_;
	my $pdu = ['STATE', $time || time, $name, $code, $value];
	my $push = Bolo::Socket->new->pusher;
	$push->connect($endpoint);
	$push->send($pdu) or die "failed to send PDU\n";
	$push->disconnect($endpoint);
}

sub shutdown
{
	my ($self) = @_;
	return unless $self->{_socket};

	if ($self->bound) {
		zmq_unbind($self->{_socket}, $_)
			for @{ $self->{_binds} };
		$self->{_binds} = [];
	}
	if ($self->connected) {
		zmq_disconnect($self->{_socket}, $_)
			for @{ $self->{_endpoints} };
		$self->{_endpoints} = [];
	}

	zmq_close($self->{_socket}) == 0
		or die "Failed to close socket: $!\n";
	$self->{_socket} = undef;
}

sub DESTROY
{
	my ($self) = @_;
	$self->shutdown;
	zmq_ctx_destroy($self->{_context});
}

1;


=head1 NAME

Bolo::Socket - the Bolo wire protocol for perl

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.


=head1 METHODS

=head2 new

=head2 timeout

=head2 timeout_us

=head2 router

=head2 dealer

=head2 pusher

=head2 puller

=head2 publisher

=head2 subscriber

=head2 bound

=head2 bind

=head2 connected

=head2 connect

=head2 disconnect

=head2 receive

=head2 send

=head2 submit_metric

    submit_metric RATE => "fqdn:test:serivce.metric1", value, $endpoint;

=head2 submit_state

    submit_state CRITICAL =>  "fqdn:test:service", "a broken thing", $endpoint;

=head2 shutdown

=head2 DESTROY

=head1 AUTHOR

Dan Molik, C<< <dan at d3fy.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<dan@d3fy.net>, or
github: https://github.com/GrayTShirt/Bolo-Socket/issues
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bolo::Socket

You can also look for information at:


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Dan Molik - GPLv3
